accepted/00000-salt-ssh-with-salt-bundle.md

```diff
@@ -0,0 +1,96 @@
+Additionally we are extending the list of working modules as Salt Bundle contains binary python modules required by some of Salt modules.
+
+This approach will help us to limit the number of supported versions of Salt codebase and drop `py26-compat-salt`, `py27-compat-salt` (as these packages are shipping `2016.11.10` and `3000.3` respectively) and all Python 2 dependencies maintenance.
+So we don't need to maintain old salt codebases, but focus only on the latest one.
+
+# Detailed design
+[design]: #detailed-design
+
+## Uyuni roster
+
+We can use Uyuni roster module to provide roster data instead of creating flat roster files with Java. PR moving the logic to Uyuni roster on Java side: https://github.com/uyuni-project/uyuni/pull/4457
+
+This change can help to avoid creating temporary roster files (https://docs.saltproject.io/en/latest/topics/ssh/roster.html) and implements caching additionally.
+
+## Pre flight script
+
+Salt can execute a shell script on the salt-ssh client before the "real salt-ssh execution" starts.
+This script downloads the salt-bundle package from the appropriate bootstrap repository and extracts the virtual environment.
+
+The possibility to pass additional parameter to the static pre flight script will be implemented to prevent creating individual pre flight script for each minion.
+The pre flight script calculates the bootstrap repository URL in a similar manner to the existing boostrap.sh script, based on the operating system and OS architecture of the system it is executed on.
+It requires one parameter from the Uyuni server: entry point (host name and port number) for different contact methods and pathes (SSH from server/proxy, direct access to the repos or with the tunnel).
+
+Pre flight script will use bootstrap repository for the distro to get the salt bundle from. No changes required for the repository.
+Pre flight script downloads the `venv-enabled-ARCH.txt` file to get full path to the salt bundle package.
+
+## Deploying Salt Bundle to the minion
+
+We don't need to install Salt Bundle package on the minion as users concider salt-ssh as an agentless contact method and it's better not to install any service on the system even if it will never been started.
+It can also help for the cases when venv-salt-minion is used on the system for any other purpose.
+Pre flight script extracts only virtual environment of the bundle from the package.
+Pre flight script is deploying it as described above.
+
+## Using proper salt codebase
+
+On deploying Salt Bundle with pre flight script, without modification salt-thin (`salt-thin` is a tarball with salt codebase created by the server to deploy salt on the salt-ssh minion: https://docs.saltproject.io/en/latest/ref/runners/all/salt.runners.thin.html) deployment, we will get two different codebases of salt deployed on the minion (one codebase is from salt bundle package for the OS+arch and the other one is from salt-master delivered with salt-thin tarball).
```

Comment:
We excpect Salt Bundle to be updated more frequently in the repo than the salt codebase installed on the master. Mostly due to the customer usecases when client tools are updating automatically and bootrap repos are regenerating after reposync and most of the customers have some delay in updating the server. In this case we can expect salt bundle codebase more fresh. We can't be 100% due to the different usecases and update policies on customer side. But in case if there is no Salt Bundle in the bootstrap repo for some reason, or pre flight script unable to reach it due to some network issues or limitations, we will add extra logging to salt-api process to show that pre flight script returns some extra data. (Currently salt is not checking the output of the script, but only fails and show error message if pre flight script returns something other than 0). But finally we will use `salt-thin` as a fallback. The main drawback here that `salt-thin` in general is limited in number of modules shipped, and in case of getting rid of `py2X-compat-salt` it will work only with Python 3. On fails of handling salt-ssh sessions with salt-thin, salt is producing error on master side showing the root cause of the issue.

Reply:
Let's discuss some additional details regarding this point.
Now we have two options:
1. Deploy Salt Bundle on the minion and fail if it's not available for some reason and do not use `salt-thin` at all. (we can just show an error message from the pre flight script)
2. Deploy Salt Bundle on the minion if possible (do not fail if not) and use `salt-thin` as a fall back method and get the error only if `salt-thin` fails.

With option 1 we can cause the issues for some of the customers. Here are the possible cases:
- The customer disabled auto bootsrap repo creation for some reason (it's mostly possible on migrating from previous version) and bootstrap repo was not updated with Salt Bundle at all. Bootstrapping new clients and handling of the sessions for already registered salt-ssh system for the OS will not be possible until bootstrap repo recreated manually.
- The customer is using salt-ssh with not supported OS and have no bootstrap repo for it at all.

The main possible drawback of the 2nd option: it's possible that some event can be handled different ways with Salt Bundle and `salt-thin` as the last one is limited in the functionality and is not shipped with all required modules inside. Additionally it builds with the salt codebase of the master, but not from the client tools (we expect client tools should be updated more frequently). As @moio pointed it could add some headache on debugging salt-ssh issues the customers have.

@meaksh @agraul @SchoolGuy not sure if I listed all of the cases we mentioned on the call.

The possible solution from me: we can use 1st option as default, but make it configurable to allow to switch to 2nd if the customer really need it for some reason.

Reply:
I think you cover the gist of what we discussed. I would like to point out that "customer using salt-ssh with not supported OS and have no bootstrap repo for it at all" is not only a potential problem for SUMA customers but also for uyuni contributors who want to enable new operating systems. While the former group is not supposed to use unsupported operating systems, we encourage the latter group to do exactly that. 

To me it makes sense to not have an automatic fallback but a manual "use salt-thin" configuration option, which we declare unsupported in SUMA.

Reply:
> To me it makes sense to not have an automatic fallback but a manual "use salt-thin" configuration option, which we declare unsupported in SUMA.

To me this sounds like a fair balancing act to start with. We can always change the policy and make that fully supported later, in case we observe a high volume of requests because of conditions we cannot predict today.